### PR TITLE
Require API endpoints to declare their access

### PR DESCRIPTION
**Pull request by spaze**
The `Api` base presenter authenticated nothing and left each endpoint to guard itself, so the module was default-allow: a newly added `Api` presenter would ship publicly reachable unless its author remembered to add a check. Both endpoints today are correct (`Certificates` authenticates with a shared secret, `Company` is deliberately public), so nothing is exposed, but relying on the author to remember is fragile as the API grows.

Now the base denies any endpoint that has not declared its access, so a new one is closed until its author consciously marks it public or authenticated. The marker is an attribute implementing a shared interface, so adding another kind of access later needs no change to the check, and a test fails if any endpoint declares none.

### CODE REVIEW

File: app/src/Presentation/Api/BasePresenter.php [hunk 1]
```diff
@@ -28,6 +30,22 @@ protected function startup(): void
 	{
 		parent::startup();
 		$this->robots->setHeader([RobotsRule::NoIndex, RobotsRule::NoFollow]);
+		$this->checkAccessAttribute();
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/src/Presentation/Api/BasePresenter.php]: Made `Api\BasePresenter::startup()` final and added an overridable `startupApi()` hook.

File: app/src/Presentation/Api/BasePresenter.php [hunk 2]
```diff
@@ -24,10 +26,35 @@ public function injectRobots(Robots $robots): void
...
+	 * namespace (public or authenticated).
+	 *
+	 * @throws BadRequestException
+	 */
+	private function checkAccessAttribute(): void
+	{
+		if (!ApiEndpointAccess::isDeclared(static::class)) {
+			throw new BadRequestException(sprintf('%s does not declare its access; add one of the attributes from the MichalSpacekCz\\Api namespace', static::class), 403);
+		}
 	}
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/src/Presentation/Api/BasePresenter.php:59]: Enforcing "exactly one" now.

File: app/src/Presentation/Api/BasePresenter.php [hunk 3]
```diff
@@ -24,10 +27,34 @@ public function injectRobots(Robots $robots): void
...
+	/**
+	 * Require exactly one access attribute from the MichalSpacekCz\Api namespace, so a newly added presenter
+	 * is closed by default (none declared) and a contradictory declaration (more than one) is rejected.
+	 *
+	 * @throws BadRequestException
+	 */
+	private function checkAccessAttribute(): void
+	{
+		if (!ApiEndpointAccess::isDeclared($this)) {
+			throw new BadRequestException(sprintf('%s must declare exactly one access attribute from the MichalSpacekCz\\Api namespace', static::class), IResponse::S403_Forbidden);
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/src/Presentation/Api/BasePresenter.php]: Changed the text.

File: app/src/Api/AuthenticatedApiEndpoint.php [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?php
+declare(strict_types = 1);
+
+namespace MichalSpacekCz\Api;
+
+use Attribute;
+
+/**
+ * Marks an API presenter that requires the caller to authenticate. The presenter performs the check
+ * itself (e.g. the shared-secret credential check in Certificates' startup()); this attribute only
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/src/Api/AuthenticatedApiEndpoint.php]: Changed the text.

File: app/tests/Api/ApiEndpointAccessTest.phpt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		}
+		$found = [];
+		foreach ($files as $file) {
+			$class = 'MichalSpacekCz\\Presentation\\Api\\' . basename(dirname($file)) . '\\' . basename($file, '.php');
+			if (class_exists($class) && is_subclass_of($class, BasePresenter::class)) {
+				$found[] = $class;
+			}
+		}
+		return $found;
+	}
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/tests/Api/ApiEndpointAccessTest.phpt]: Now scanning recursively.

File: app/src/Presentation/Api/BasePresenter.php [hunk 6]
```diff
@@ -24,10 +27,34 @@ public function injectRobots(Robots $robots): void
...
+	 */
+	protected function startupApi(): void
+	{
+	}
+
+
+	/**
+	 * Require exactly one access attribute from the MichalSpacekCz\Api namespace, so a newly added presenter
+	 * is closed by default (none declared) and a contradictory declaration (more than one) is rejected.
+	 *
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/src/Presentation/Api/BasePresenter.php]: Changed the text.

File: app/src/Api/Endpoint/EndpointRequiresAuthentication.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+declare(strict_types = 1);
+
+namespace MichalSpacekCz\Api\Endpoint;
+
+use Attribute;
+
+/**
+ * Marks an API endpoint that requires the caller to authenticate. The endpoint performs the check itself
+ * (e.g. the shared-secret credential check in Certificates::startupApi()); this attribute only declares the
+ * requirement, so the base presenter lets the endpoint run rather than denying it by default.
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/src/Api/Endpoint/EndpointRequiresAuthentication.php]: Dropped the class reference, it drifts faster than they did in the whole Fast & Furious saga.

File: app/tests/Api/Endpoint/EndpointAccessTest.phpt [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+			assert($file instanceof SplFileInfo);
+			if (!$file->isFile() || !str_ends_with($file->getPathname(), 'Presenter.php')) {
+				continue;
+			}
+			$relative = substr($file->getPathname(), strlen($apiDir) + 1, -strlen('.php'));
+			$class = 'MichalSpacekCz\\Presentation\\Api\\' . str_replace('/', '\\', $relative);
+			if (class_exists($class) && is_subclass_of($class, BasePresenter::class)) {
+				$found[] = $class;
+			}
+		}
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/tests/Api/Endpoint/EndpointAccessTest.phpt]: `realpath()` using now

File: app/tests/Api/Endpoint/EndpointAccessTest.phpt [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+			$realPath = realpath($file->getPathname());
+			if ($realPath === false) {
+				continue;
+			}
+			$relative = substr($realPath, strlen($apiDir) + 1, -strlen('.php'));
+			$class = 'MichalSpacekCz\\Presentation\\Api\\' . str_replace(DIRECTORY_SEPARATOR, '\\', $relative);
+			if (class_exists($class) && is_subclass_of($class, BasePresenter::class)) {
+				$found[] = $class;
+			}
+		}
```
[the remark this answers is not in the document]
Comment (spaze) [reply app/tests/Api/Endpoint/EndpointAccessTest.phpt:94]: Added `NotAnEndpoint` attribute to mark those error presenters.
